### PR TITLE
Add statemutability field to ABI

### PR DESCRIPTION
**Pull request by axic**
Depends on #2722. Part of #992.

### CODE REVIEW

File: Changelog.md [hunk 1]
```diff
@@ -1,6 +1,7 @@
 ### 0.4.16 (unreleased)
 
 Features:
+ * ABI: Include new field ``statemutability`` with values ``view``, ``nonpayable`` and ``payable``.
```
Comment (chriseth) [inline Changelog.md]: More specifically: json-ABI. Also, why is `pure` not a valid option? Should we name it `stateaccess`? Or `stateaccessrestrictions`?

Comment (chriseth) [reply Changelog.md]: Argh, sorry, of course because `pure` is not yet part of this PR...

Comment (axic) [reply Changelog.md]: Going to add ``pure`` in a separate PR.

File: libsolidity/ast/ASTEnums.h [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+namespace dev
+{
+namespace solidity
+{
+
+// How a function can mutate the EVM state.
+enum class StateMutability { View, NonPayable, Payable };
+
+inline std::string stateMutabilityToString(StateMutability const& _stateMutability)
```
Comment (chriseth) [inline libsolidity/ast/ASTEnums.h]: C++ has nicely-usable overloads, we can call this just `toString`.

File: libsolidity/ast/Types.h [hunk 3]
```diff
@@ -985,6 +982,7 @@ class FunctionType: public Type
 	/// @returns true if the ABI is used for this call (only meaningful for external calls)
 	bool isBareCall() const;
 	Kind const& kind() const { return m_kind; }
+	StateMutability const& stateMutability() const { return m_stateMutability; }
```
Comment (chriseth) [inline libsolidity/ast/Types.h]: It is probably better to return by value.

File: libsolidity/ast/ASTEnums.h (hunk 2 above)
Comment (axic) [reply libsolidity/ast/ASTEnums.h]: I think this part of the review belongs to #2722.

**Review by chriseth — APPROVED** [review]
Only minor requests, can also be merged as is, if needed.

File: libsolidity/ast/Types.h (hunk 3 above)
Comment (axic) [reply libsolidity/ast/Types.h]: Addressed this in the real PR.

File: libsolidity/ast/ASTEnums.h (hunk 2 above)
Comment (axic) [inline libsolidity/ast/ASTEnums.h]: Addressed this in the real PR.

**Comment by axic** [comment]
Need to update tests.

File: test/libsolidity/SolidityABIJSON.cpp [hunk 4]
```diff
@@ -548,12 +561,44 @@ BOOST_AUTO_TEST_CASE(constructor_abi)
 			}
 		],
 		"payable": false,
+		"statemutability" :"nonpayable",
```
Comment (axic) [inline test/libsolidity/SolidityABIJSON.cpp]: Wrong space around colon.

File: test/libsolidity/SolidityABIJSON.cpp [hunk 5]
```diff
@@ -548,12 +561,44 @@ BOOST_AUTO_TEST_CASE(constructor_abi)
...
+				"name": "param2",
+				"type": "address"
+			},
+			{
+				"name": "param3",
+				"type": "bool"
+			}
+		],
+		"payable": true,
+		"statemutability" :"payable",
```
Comment (axic) [inline test/libsolidity/SolidityABIJSON.cpp]: Wrong space around colon.

**Comment by axic** [comment]
@chriseth ready to merge

**Comment by axic** [comment]
@chriseth check this? :)

**Pull request merged**